polling_stations/apps/data_importers/data_types.py

```diff
@@ -0,0 +1,291 @@
+"""
+Data type classes used by base importers
+"""
+
+import abc
+import logging
+from collections import namedtuple
+
+from django.db import connection
+
+from addressbase.models import get_uprn_hash_table, UprnToCouncil
+from pollingstations.models import PollingDistrict, PollingStation
+from uk_geo_utils.helpers import Postcode
+
+Station = namedtuple(
+    "Station",
+    [
+        "council",
+        "internal_council_id",
+        "postcode",
+        "address",
+        "location",
+        "polling_district_id",
+    ],
+)
+
+
+District = namedtuple(
+    "District",
+    [
+        "name",
+        "council",
+        "internal_council_id",
+        "extra_id",
+        "area",
+        "polling_station_id",
+    ],
+)
+
+
+class CustomSet(metaclass=abc.ABCMeta):
+    def __init__(self):
+        self.elements = set()
+
+    def add(self, element):
+        self.elements.add(self.build_namedtuple(element))
+
+    @abc.abstractmethod
+    def build_namedtuple(self, element):
+        pass
+
+
+class AssignPollingStationsMixin(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def get_polling_station_lookup(self):
+        pass
+
+    @property
+    def council_id(self):  # TODO Deal with old_to_new council_ids map
+        for e in self.elements:
+            if isinstance(e, dict):
+                return e["council"].council_id
+            else:
+                return e.council.council_id
+
+    def update_uprn_to_council_model(self, polling_station_lookup=None):
+        if not polling_station_lookup:
+            polling_station_lookup = self.get_polling_station_lookup()
+
+        uprns_in_council = UprnToCouncil.objects.filter(lad=self.council_id)
+        seen = set()
+        for polling_station_id, uprns in polling_station_lookup.items():
+            uprns_in_council.filter(uprn__in=uprns).update(
+                polling_station_id=polling_station_id
+            )
+            if self.__class__.__name__ == "DistrictSet":
```

Comment:
If there are multiple classes that extend the same base class, it is more normal for the child class to override the behaviour of the parent class with a special behaviour than for the parent to implement special behaviour for the child.